### PR TITLE
#1141 - Garden connection import/export

### PR DESCRIPTION
**Pull request by jlrpnbbngtn**
Closes #1141

### CODE REVIEW

File: src/app/beer_garden/api/http/handlers/v1/garden.py [hunk 1]
```diff
@@ -41,7 +51,7 @@ async def get(self, garden_name):
         """
         garden = self.get_or_raise(Garden, GARDEN_READ, name=garden_name)
 
-        response = GardenSchema(strict=True).dumps(garden).data
+        response = self.validate_or_raise(GardenSchema, garden, method="dumps")
```
Comment (scott-taubman) [inline src/app/beer_garden/api/http/handlers/v1/garden.py]: I made an earlier comment indicating that I wasn't sure when you would use the "dump" option on this method. 
This is actually a good example of what I was trying to explain there.

You only want to use the `validate_or_raise` in situations where what you're validating is user input.  The thing being validated here is a valid garden that we've already retrieved based on the user input.  Meaning the user's input has already been determined to be good.  If the dump / validation fails at this point, it's because of an unexpected error on our end, not the users, meaning letting the code blow up and throw a 500 (Internal Server Error) is correct.

It looks like there are a few other instances where you're doing this as well so we'll want to change those as well.  I'll comment in places where I spot them just to keep track.

File: src/app/test/db/mongo/models/garden_model_test.py [hunk 2]
```diff
@@ -225,8 +225,8 @@ def test_local_garden_save_succeeds_with_empty_conn_params(self):
         "conn_parm",
         (
             pytest.lazy_fixture("bad_conn_params"),
-            # pytest.lazy_fixture("bad_conn_params_with_partial_good"),
-            # pytest.lazy_fixture("bad_conn_params_with_full_good"),
+            pytest.lazy_fixture("bad_conn_params_with_partial_good"),
```
Comment (jlrpnbbngtn) [inline src/app/test/db/mongo/models/garden_model_test.py:228]: These were mistakenly left commented from before.

File: src/app/beer_garden/api/http/base_handler.py [hunk 3]
```diff
@@ -250,3 +251,29 @@ def schema_validated_body(self, schema: Type[Schema]) -> dict:
             return schema(strict=True).load(self.request_body).data
         except MarshmallowValidationError:
             raise BadRequest
+
+    def validate_or_raise(
```
Comment (jlrpnbbngtn) [inline src/app/beer_garden/api/http/base_handler.py]: This is similar, in spirit, to `schema_validated_body` but adds some useful features:

1. It is more general. It works on more than just the request body which was helpful in the new endpoint code (**D***on't* **R***epeat* **Y***ourself*).
2. It raises a specific HTTP exeception: `ValidationError` defined in `beer_garden.api.http.exceptions`. Moreover, it preserves passed messages for the `reason`, which is expected to be useful to clients. In particular, the UI code, but also the `http_client` used in unit tests.

With that said, there are probably uses of this function where it isn't strictly necessary. Example below...

File: src/app/beer_garden/api/http/base_handler.py [hunk 4]
```diff
@@ -198,7 +199,7 @@ def write_error(self, status_code, **kwargs):
                 message = error_dict.get("message", getattr(e, "message", str(e)))
                 code = error_dict.get("status_code", 500)
             elif issubclass(typ3, BaseHTTPError):
-                message = typ3.reason
+                message = self._reason
```
Comment (jlrpnbbngtn) [inline src/app/beer_garden/api/http/base_handler.py:201]: `self._reason` is set in the parent correctly before this is called.

Previously, `typ3.reason` was used and *that*, for unknown reasons, gets overwritten to the class variable value at some point prior to this call. It went unnoticed before because the `reason` field probably wasn't specifically set in any thrown exception. This code preserves the passed-in message during exception instantiation (when it exists) and falls back to the class value if no `reason` argument is provided to the exception's `__init__`.

File: src/app/beer_garden/api/http/base_handler.py [hunk 5]
```diff
@@ -210,7 +211,7 @@ def write_error(self, status_code, **kwargs):
         )
 
         self.set_header("Content-Type", "application/json; charset=UTF-8")
-        self.set_status(code)
+        self.set_status(code, reason=message)
```
Comment (jlrpnbbngtn) [inline src/app/beer_garden/api/http/base_handler.py:213]: This allows clients, such as tornado's `http_client` (used in unit tests), access to the `message` value. Otherwise, the `http_client` writes its own `reason` based on the status code.

File: src/app/beer_garden/api/http/handlers/v1/garden.py [hunk 6]
```diff
@@ -68,7 +78,9 @@ async def delete(self, garden_name):
         """
         garden = self.get_or_raise(Garden, GARDEN_DELETE, name=garden_name)
 
-        await self.client(Operation(operation_type="GARDEN_DELETE", args=[garden.name]))
+        await self.client(
+            BrewtilsOperation(operation_type="GARDEN_DELETE", args=[garden.name])
```
Comment (jlrpnbbngtn) [inline src/app/beer_garden/api/http/handlers/v1/garden.py:76]: `Operation` becomes `BrewtilsOperation` for consistency with what we've been doing elsewhere.

File: src/app/beer_garden/garden.py [hunk 7]
```diff
@@ -186,29 +186,34 @@ def get_connection_defaults():
...
-            defaults[key]
-            if (
-                (defaults[key] is not None and (isinstance(defaults[key], bool)))
-                or defaults[key]
-            )
-            else sensible_defaults[key]
-        )
-        for key in config_map
-    }
+    new_defaults = {}
```
Comment (jlrpnbbngtn) [inline src/app/beer_garden/garden.py:199]: This was cleaned up from the previous version. Semantically the same, but there were some edge cases the other version would barf on.

File: src/app/test/db/mongo/models/models_test.py [hunk 8]
```diff
@@ -6,7 +6,7 @@
 from brewtils.errors import ModelValidationError, RequestStatusTransitionError
 from brewtils.schemas import RequestTemplateSchema
 from mock import Mock
-from mongoengine import ValidationError, connect
+from mongoengine import NotUniqueError, ValidationError, connect
```
Comment (jlrpnbbngtn) [inline src/app/test/db/mongo/models/models_test.py]: No idea how this got omitted despite what appeared to be correct rebasing. At any rate, ignore. It's what a previously approved PR added.

File: src/app/beer_garden/api/http/handlers/v1/garden.py [hunk 9]
```diff
@@ -182,8 +196,9 @@ async def get(self):
           - Garden
         """
         permitted_gardens: QuerySet = self.permissioned_queryset(Garden, GARDEN_READ)
-
-        response = GardenSchema(strict=True, many=True).dumps(permitted_gardens).data
+        response = self.validate_or_raise(
+            GardenSchema, permitted_gardens, method="dumps", many=True
+        )
```
Comment (jlrpnbbngtn) [inline src/app/beer_garden/api/http/handlers/v1/garden.py]: This, for example, could be replaced by

```python
response = GardenSchema(strict=True, many=True).dumps(permitted_gardens).data
```
because we can be sure that the result of `permissioned_queryset` won't throw a validation error. But that's really no shorter. Perhaps the function can be aliased to just plain `validate` to make it clearer...

File: src/app/beer_garden/api/http/base_handler.py [hunk 10]
```diff
@@ -250,3 +251,29 @@ def schema_validated_body(self, schema: Type[Schema]) -> dict:
...
+        self, schema: Type[Schema], arg: Any, method: str = "loads", many: bool = False
+    ):
+        """Apply a provided schema to an argument and raise an exception if it fails
+        to validate.
+
+        Args:
+            schema: A schema derived from a marshmallow Schema that the argment
+                will be validated against
+            arg: The data to validate
+            method: The method to call; one of `load`, `loads`, `dump` or `dumps`
```
Comment (scott-taubman) [inline src/app/beer_garden/api/http/base_handler.py]: This part seems a little odd to me.  The idea that you'd call a method called "validate..." and then provide a method for it to call in addition to the validation.  It feels like an odd pattern, but also, I'm trying to think of the scenario where you would use the dump option here.  If you're serializing an object (calling dump) it's somewhat implied that the object you're dealing with was retrieved from some internal call, not user input, since user input comes in the form a JSON (or a dict, if the JSON has already been converted).

File: src/app/beer_garden/api/http/handlers/v1/garden.py (hunk 6 above)
Comment (scott-taubman) [reply src/app/beer_garden/api/http/handlers/v1/garden.py:76]: There is actually no beer garden model named Operation, so there's nothing to deconflict here.  The reason I started doing it elsewhere was because there were conflicting model names.  You can carry on with this if you'd like, but if you do we should go through later and rename this everywhere.

File: src/app/beer_garden/api/http/handlers/v1/garden.py (hunk 9 above)
Comment (scott-taubman) [reply src/app/beer_garden/api/http/handlers/v1/garden.py]: This is another instance where I don't think you want to use the `validate_or_raise`.  If this dump fails to validate, it's because of an internal error, not a user input issue.

File: src/app/beer_garden/api/http/handlers/v1/garden.py [hunk 11]
```diff
@@ -285,3 +302,84 @@ async def patch(self):
...
+            schema:
+              $ref: '#/definitions/GardenConnectionConfiguration'
+          404:
+            $ref: '#/definitions/404Error'
+          50x:
+            $ref: '#/definitions/50xError'
+        tags:
+          - Garden
+        """
+        garden: BrewtilsGarden = self.get_or_raise(
```
Comment (scott-taubman) [inline src/app/beer_garden/api/http/handlers/v1/garden.py]: `get_or_raise` actually returns a beer garden model object, not a brewtils model.  Which you'd know if I'd put type hinting on the method.  Sorry about that.

File: src/app/beer_garden/api/http/handlers/v1/garden.py [hunk 12]
```diff
@@ -285,3 +302,84 @@ async def patch(self):
...
+            Garden, GARDEN_READ, name=garden_name
+        )
+        connection_params = garden.connection_params
+
+        response = self.validate_or_raise(
```
Comment (scott-taubman) [inline src/app/beer_garden/api/http/handlers/v1/garden.py]: Another one...

File: src/app/beer_garden/api/http/handlers/v1/garden.py [hunk 13]
```diff
@@ -285,3 +302,84 @@ async def patch(self):
...
+            description: The name of the Garden whose connection will be updated
+            type: string
+          - name: config
+            in: body
+            required: true
+            description: The new connection parameters
+            schema:
+              $ref: '#/definitions/GardenConnectionConfiguration'
+        responses:
+          201:
```
Comment (scott-taubman) [inline src/app/beer_garden/api/http/handlers/v1/garden.py]: 201 is "created", which I don't think I typically see used for PATCH operations.  That's effectively what this is, even though we've exposed it via this import endpoint.  Which actually raises it's own question about why this isn't just a PATCH on the garden.  Maybe it should be?  I'll make a separate comment to discuss that.  For what's here currently, since you're returning a body, 200 is probably appropriate.

File: src/app/beer_garden/api/http/handlers/v1/garden.py [hunk 14]
```diff
@@ -285,3 +302,84 @@ async def patch(self):
...
+            schema:
+              $ref: '#/definitions/Garden'
+          400:
+            $ref: '#/definitions/400Error'
+          50x:
+            $ref: '#/definitions/50xError'
+        tags:
+          - Garden
+        """
+        garden_to_update: BrewtilsGarden = self.get_or_raise(
```
Comment (scott-taubman) [inline src/app/beer_garden/api/http/handlers/v1/garden.py]: See earlier comment about what `get_or_raise` returns.  I apologize again.

File: src/app/beer_garden/api/http/handlers/v1/garden.py [hunk 15]
```diff
@@ -285,3 +302,84 @@ async def patch(self):
...
+            Garden, GARDEN_READ, name=garden_name
+        )
+        garden_to_update.connection_params = self.validate_or_raise(
+            GardenConnectionsParamsSchema, self.request.decoded_body, method="loads"
+        )
+
+        response = self.validate_or_raise(
```
Comment (scott-taubman) [inline src/app/beer_garden/api/http/handlers/v1/garden.py]: Another one...

File: src/app/test/api/http/unit/conftest.py [hunk 16]
```diff
@@ -44,6 +49,8 @@
         (r"/api/v1/forward/?", ForwardAPI),
         (r"/api/v1/gardens/?", GardenListAPI),
         (r"/api/v1/gardens/(.*)/?", GardenAPI),
+        (r"/api/v1/export/connections/(\w+)/?", GardenConnectionParametersExportAPI),
+        (r"/api/v1/import/connections/(\w+)/?", GardenConnectionParametersImportAPI),
```
Comment (scott-taubman) [inline src/app/test/api/http/unit/conftest.py]: Picking an appropriate, but still a little random, spot to start the "is this just a PATCH?" conversation I alluded to earlier.

I can see that you somewhat followed the import / export pattern that you previously established with the job import / export.  There's a key distinction here though in that you've made these endpoints operate on a single garden, as where the job endpoints operate on multiple jobs.

By operating on just one garden, it's not really clear to me that there's a meaningful distinction between these new endpoints and the existing ones.  That is, `/api/v1/export/connections/somegarden` is really just returning a subset of the same data you could already get via `/api/v1/gardens/somegarden`.  And `/api/v1/import/connections/somegarden` is just a more limited version of the existing `PATCH` you could already do.

A couple of questions that I think we need to answer:
* Is there functionality / value added by the single garden version of these import / export endpoints that I'm not seeing?
* Whether there is or not, does it make more sense have these operate on multiple gardens similar to how the job import / export works?  I'm inclined to say yes, but if you went down this current path for a specific reason then we certainly can discuss it.

File: src/app/beer_garden/api/http/handlers/v1/garden.py [hunk 17]
```diff
@@ -1,14 +1,24 @@
 # -*- coding: utf-8 -*-
+import logging
+
 from brewtils.errors import ModelValidationError
-from brewtils.models import Operation
+from brewtils.models import Garden as BrewtilsGarden
+from brewtils.models import Operation as BrewtilsOperation
 from brewtils.schema_parser import SchemaParser
+from marshmallow.exceptions import ValidationError as MarshmallowValidationError
```
Comment (scott-taubman) [inline src/app/beer_garden/api/http/handlers/v1/garden.py]: This and the other `ValidationError` don't appear to get used.

File: src/app/beer_garden/api/http/handlers/v1/garden.py (hunk 6 above)
Comment (jlrpnbbngtn) [reply src/app/beer_garden/api/http/handlers/v1/garden.py:76]: I like having the *Brewtils* prefix if for no other reason than I can later know exactly where to kill it with fire when the time comes ;)

File: src/app/test/api/http/unit/conftest.py (hunk 16 above)
Comment (jlrpnbbngtn) [reply src/app/test/api/http/unit/conftest.py]: Point taken and I noticed that the new endpoints weren't adding very much to the mix. But:
* The reason a new endpoint was added -- vice reusing the patch code -- was because a colleague who has since moved on decided that `.../api/v1/export/...` and `.../api/v1/import/...` constituted good API design and that import/export command should live there. I followed suit both originally and in what I expect is the first in a series of import/export endpoints. Regardless, having all export functions under an `.../export/...` path *feels* like a better design than having some export functions in one place, others in another. Consistency etc.
* As far as the UI goes, it needs a mechanism to export and import only a single garden. It was agreed that the ultimate buttons that will make it into the UI will be on the specific garden's config page.
* Alternately (or additionally), there *could* be an "export all garden connection configs" button, but I recall that you thought at that time of the discussion that the re-import of a list of configs would be problematic. I think the argument was that the name of a garden is the only discriminator. If the UI always gets back a list of garden configs when exporting or updating, that's the same problem as far as I can tell.

File: src/app/beer_garden/api/http/handlers/v1/garden.py (hunk 1 above)
Comment (jlrpnbbngtn) [reply src/app/beer_garden/api/http/handlers/v1/garden.py]: Fair enough. Working on it.

File: src/app/beer_garden/api/http/base_handler.py (hunk 10 above)
Comment (jlrpnbbngtn) [reply src/app/beer_garden/api/http/base_handler.py]: Converted to the more typical "from_string" arg.

File: src/app/beer_garden/api/http/base_handler.py [hunk 18]
```diff
@@ -269,11 +273,39 @@ def validate_or_raise(
...
+        schema = schema(strict=True, many=many)
 
         try:
-            return getattr(schema, method)(arg).data
+            return schema.loads(arg).data if from_string else schema.load(arg).data
         except MarshmallowValidationError as mmve:
-            raise ValidationError(reason=str(mmve))
+            raise BadRequest(reason=str(mmve))
+
+    def format_response(
```
Comment (jlrpnbbngtn) [inline src/app/beer_garden/api/http/base_handler.py:285]: I'm not married to this name.

File: src/app/beer_garden/api/http/handlers/v1/garden.py [hunk 19]
```diff
@@ -152,9 +146,10 @@ async def patch(self, garden_name):
                 )
             elif operation == "config":
                 garden_to_update = self.validate_or_raise(
-                    GardenSchema, op.value, method="load"
+                    GardenSchema, op.value, from_string=False
```
Comment (jlrpnbbngtn) [inline src/app/beer_garden/api/http/handlers/v1/garden.py:149]: If a bad garden JSON is passed, it will blow up here.

File: src/app/beer_garden/api/http/handlers/v1/garden.py [hunk 20]
```diff
@@ -152,9 +146,10 @@ async def patch(self, garden_name):
...
                 )
                 garden_to_update.id = garden.id
+
                 response = await self.client(
```
Comment (jlrpnbbngtn) [inline src/app/beer_garden/api/http/handlers/v1/garden.py:155]: And it shouldn't blow up here.

File: src/app/beer_garden/events/handlers.py [hunk 21]
```diff
@@ -34,19 +34,22 @@ def garden_callbacks(event: Event) -> None:
...
-        beer_garden.plugin.handle_event,
-        beer_garden.requests.handle_event,
-        beer_garden.router.handle_event,
-        beer_garden.systems.handle_event,
-        beer_garden.scheduler.handle_event,
-        beer_garden.log.handle_event,
-        beer_garden.files.handle_event,
-        beer_garden.local_plugins.manager.handle_event,
+    for handler, handler_tag in [
+        (beer_garden.application.handle_event, "Application"),
```
Comment (jlrpnbbngtn) [inline src/app/beer_garden/events/handlers.py:38]: Added a tag to make the error messages more informative.

File: src/app/beer_garden/api/http/base_handler.py (hunk 10 above)
Comment (scott-taubman) [reply src/app/beer_garden/api/http/base_handler.py]: Yeah, I like that better.  Maybe `load_or_raise` would be a better name for this now?  Otherwise it still feels weird to call a method named "validate" that is in fact doing a load and returning an object.

Also, it looks like you docstring needs to be updated to reflect the changes.

File: src/app/beer_garden/api/http/base_handler.py (hunk 18 above)
Comment (scott-taubman) [reply src/app/beer_garden/api/http/base_handler.py:285]: This name seems okay.  This had me thinking about a previous conversation we had somewhere where I think we you trying to detect whether `many=True` should be set or not and it didn't feel appropriate to me in that instance.  Here it feels like that concept could make sense if you wanted to apply it.  Basically, `format_response` could have `many` be totally optional and if it's not set, you set it to true if the received `arg` is iterable or some such thing.  Definitely not necessary, but I remembered you attempting to do this somewhere else so I figured I'd bring it up as an option.

File: src/app/beer_garden/api/http/handlers/v1/garden.py [hunk 22]
```diff
@@ -210,12 +218,14 @@ async def post(self):
         tags:
           - Garden
         """
-        garden = GardenSchema(strict=True).loads(self.request.decoded_body).data
+        garden = self.validate_or_raise(
```
Comment (scott-taubman) [inline src/app/beer_garden/api/http/handlers/v1/garden.py]: This looks like it's calling the old form of `validate_or_raise` rather than your updated one (it's still taking a `method`).

File: src/app/beer_garden/api/http/handlers/v1/garden.py (hunk 19 above)
Comment (scott-taubman) [reply src/app/beer_garden/api/http/handlers/v1/garden.py:149]: A few things that came mind when looking at this again:

* One thing worth noting, which is different from your original, separate endpoint version, is that this takes a complete garden definition, rather than just the `connection_params` portion of the config,  That seems okay based on the way `GardenSchema` is currently defined, since everything but the name is considered optional.
*  The name being required when it's already provided in the path isn't great, so you might want to look at populating it based on the path parameter.
*  Name really shouldn't be a thing that you can change.  I think this endpoint will let you do it, but doing so will break a ton of stuff.  That's an existing problem so you don't have to fix it here, I just mention it because I don't want you to go to any trouble to ensure that changing the name could theoretically be done here.  I think you just take the name from the path parameter and not allow it to be set separately here.
